### ISSUE TITLE
[BUG] Bootstrapping fails for RKE2 cluster created with custom Harvester cloud provider templates

### ISSUE THREAD

**Issue by staedter** [harvester/harvester#4568]
### Describe the bug
When creating a new RKE2 cluster using custom Harvester cloud provider templates, the bootstrapping process fails. The `/var/lib/rancher/rke2/etc/config-files/cloud-provider-config` is not automatically populated, unlike when creating a cluster via the GUI.

### To Reproduce
1. Use custom Harvester cloud provider templates to create an RKE2 cluster.
2. Bootstrapping fails due to missing `/var/lib/rancher/rke2/etc/config-files/cloud-provider-config`.

### Expected behavior
The `/var/lib/rancher/rke2/etc/config-files/cloud-provider-config` should be automatically populated, similar to when creating a cluster via the GUI.

### Support bundle
- Will provide upon request.

### Environment
- Harvester ISO version: 1.2.0
- Underlying Infrastructure: On-premise cluster

### Additional context
The bootstrapping process only succeeds when I manually populate the `/var/lib/rancher/rke2/etc/config-files/cloud-provider-config` with the apropriate cloud-provider-config values, that I generate using a modified version (for convenience purposes) of the generate_addon.sh from the harvester-cloud-provider repo. This valus is then rendered into the correct path either using cloud-init write_files section or the key rkeConfig.machineSelectorConfig[0].config.cloud-provider-config in my cluster-template helm chart.

Additionally, I've noticed that the `certificate-authority-data` in the generated config is incorrect, presumably because it takes the server value from the Rancher-proxied k8s API endpoint and not the actual RKE2 server. This seems like a separate issue but is worth mentioning.

Would like to know why it's necessary to manually populate and mount the cloud-provider config, and why it doesn't happen automatically as suggested by the documentation.

**Comment by connorkuehl** [comment]
Hi @staedter, thanks for the bug report! I'm a little lost following your reproduction instructions.

What do you mean by custom Harvester cloud provider templates? Are you applying a cluster.provisioning.cattle.io/v1 CRD to your Rancher cluster to drive the creation of the RKE2 cluster?

**Comment by staedter** [comment]
Yes exactly. My cluster templates are modeled after the Rancher Master Class about this subject and Bastian Hoffmann's example Repo but with some fixes/improvements.

I will provide all the necessary step to reproduce when I am back at my desk tommorow, then I can share my setup and an example helm release.

**Comment by staedter** [comment]
Hi, I have put a sanitizied version of our cluster-templates on github. https://github.com/staedter/example-cluster-templates

There should be everything in there to reproduce the issue using an available Harvester Cluster. I hope the information in the README.md is sufficient to get you up and running. But if it is not, I will gladly provide more detailed steps.

Thanks for looking into the issue.

**Comment by connorkuehl** [comment]
Thank you for the detailed reproducer! Just to make sure I understand the scenario correctly:

When creating a clusters.provisioning object _without_ the [.spec.rkeConfig.machineSelectorConfig.["config"]["cloud-provider-config"]](https://github.com/staedter/example-cluster-templates/blob/675efaef30a0db547bc3893261da426edd903a2a/example-cluster-template/values.yaml#L50C35-L50C35) field set (and _without_ the [userdata workaround](https://github.com/staedter/example-cluster-templates/blob/675efaef30a0db547bc3893261da426edd903a2a/example-cluster-template/values.yaml#L19)):

Expected behavior: Rancher automatically renders the proper kubeconfig and propagates it into the guest VM

Actual behavior: file is not rendered at all

Is my understanding correct?

**Comment by staedter** [comment]
Yes correct. I have to use the workaround, otherwise the harvester-cloud-provider pod will be stuck pending because the cloud-provider-config file is nowhere to be found on the nodes.

**Comment by connorkuehl** [comment]
Hey @staedter, I've been looking into this and asking around. The reason why the experience is so different for applying the CRD vs using the Rancher UI is because the Rancher UI's setup flow asks for the necessary information to render a kubeconfig that is scoped appropriately in terms of service accounts, namespaces, etc, as an input to the clusters.provisioning object.

Since the Rancher UI isn't involved in the flow you're after, you'll have to continue using the `generate_addon.sh` script and supplying it either via your userdata workaround or the cloud-provider-config key in the clusters.provisioning spec.

**Comment by staedter** [comment]
Hey @connorkuehl, thanks for getting back to me.

Ok at least I wasn't doing anything wrong then. I guess than the "magic" happens somewhere in the rancher api or the frontend? Any way thanks for asking around.

The main problem I have with the workaround is that it has slowed the deployment of new pools noticeably, probably because of the giant b64 bodies that have to be passed around to rancher and harvester.

Do you know if it is possible to gzip those strings or maybe just create a secret with the kubeconfig and maybe just reference that in the CR?

**Comment by connorkuehl** [comment]
> Ok at least I wasn't doing anything wrong then.

Right.

>  I guess than the "magic" happens somewhere in the rancher api or the frontend? Any way thanks for asking around.

Yep! The Rancher frontend does the equivalent of the `generate_addon.sh`.

> just create a secret with the kubeconfig and maybe just reference that in the CR?

Yeah, this one! If your value for the `cloud-provider-config` key starts with `secret://` then Rancher will look for a named secret for the kubeconfig.

For example:

```
    machineSelectorConfig:
      - config:
          cloud-provider-config: secret://fleet-default:harvesterconfig6mkkp
          cloud-provider-name: harvester
```

The kubeconfig contents should be in that secret under the `credential` key. As you can see, the format is `secret://` followed by the namespace where the secret is and then a colon with the secret name.

**Comment by staedter** [comment]
Hi @connorkuehl

I was preoccupied with other aspect of our new cluster and because the user-data workaround is working, I postponed following up on your suggestion.

I was finally able to test it, but I think there I am doing something wrong.
I have created a new secret called "cloudconfigsecret-test" in the "fleet-default" namespace in my rancher cluster and have pasted the kubeconfig output from the generate_addon.sh as a base64-string into the secret as the value of  the key "credential" like this:

![image](https://github.com/harvester/harvester/assets/46670623/b585b258-6035-4a13-935e-77883c82c6f7)

And I have also changed my override-values to this:

![image](https://github.com/harvester/harvester/assets/46670623/c19e75e2-f1bf-45e4-996d-e0f36e2516bb)

But on the node the file is not present and thus the harvester-cloud-provider will not come up and finish the cluster bootstrapping:
![image](https://github.com/harvester/harvester/assets/46670623/ceca6292-be44-4367-b142-ea48300c6fc9)

Do you maybe see what I am doing wrong, this time?

**Comment by johnwc** [comment]
@connorkuehl can this please be reopened as a feature request. We need to be able to deploy clusters from configs, and not solely via the UI.

**Comment by nurtext** [comment]
@johnwc @connorkuehl I second this.

**Comment by staedter** [comment]
We are also now seeing problems updating our clusters with the workaround, probably because if I understood it correctly, the cloud provider credentials can now expire for the harvester cloud provider in rancher. (see this current Issue https://github.com/rancher/rancher/issues/44912). Could this be the reason our workflow broke suddenly?

I think we should work together to find a more stable solution for managing downstream cluster with in an IaC workflow.

**Comment by connorkuehl** [comment]
Hi everyone, my participation in the Harvester project ended a few months ago. Please file a new ticket so that active team members can see this, or link this one in the Rancher Users slack and ask them to re-open it.

**Comment by staedter** [comment]
Thanks for the heads-up. I think our problem was either user error (i.e., I fumbled something while using our workaround) or a compatibility issue between Rancher 2.7.9 and Harvester 1.3.1.

After upgrading Rancher to 2.8.4, it seems to work. However, I think I will try to get the new Rancher people involved again, to maybe find a really supported way to use the cluster-templates as IaC with gitops in fleet, which is our goal.

Thanks for all the help, @connorkuehl, and good luck with your future endeavors.

**Comment by johnwc** [comment]
@w13915984028  can you help get this issue reassigned to someone currently active on the team?

**Comment by johnwc** [comment]
@Yu-Jack why did you close this? This is still a major issue when trying to automate cluster creation and management via templates. 

@w13915984028 what can we do to get this issue more attention? We really need to have rancher generate this secret on-demand when a cluster is created and has the cloud-provider set to harvester, so we are not having to manually run a script prior to deploying every cluster. (Maybe for other cloud providers as well if they too require it.)

@staedter I modified the `generate_addon.sh` quite a bit to automate creating a secret in rancher, that then can be used in the template values/questions.yaml (using `type: reference[secret]`) So, the only thing that needs to be done, is run the script using a kubeconfig for rancher to generate the secret, then go into rancher and deploy a cluster via template choosing the secret. How can I share it with you and @bcdurden? BTW, I figured out where the correct certificate is located for building the kubeconfig for the secret.

**Comment by Yu-Jack** [comment]
HI @johnwc,

Actually, I don't close it. It was closed by other person before. If you need, we're welcome you to reopen it. 

![Image](https://github.com/user-attachments/assets/47194579-31a2-4dbf-bfd2-e3415c20801f)

**Comment by johnwc** [comment]
@Yu-Jack I'm confused, what is the "moving" that is happening here?

![Image](https://github.com/user-attachments/assets/768c78bb-60e4-4921-a105-0b1e1f455b48)

**Comment by Yu-Jack** [comment]
@johnwc That's a status syncing between our internal project status. Just ignore it.

**Comment by w13915984028** [comment]
@johnwc 

Sorry the issue has lasted for a long time.

Could you share a latest example or write down the detailed steps to reproduce the issue, your observation/suggestions..., then we continue the investigation from your new output, thanks.
